apm-agent-core/src/main/java/co/elastic/apm/agent/report/IntakeV2ReportingEventHandler.java

```diff
@@ -164,27 +166,52 @@ private void handleEvent(ReportingEvent event, long sequence, boolean endOfBatch
             return;
         }
         processorEventHandler.onEvent(event, sequence, endOfBatch);
-        if (connection == null) {
-            if (gracePeriodEnd > System.currentTimeMillis()) {
-                // back off because there are connection issues with the apm server
-                dropped++;
-                return;
-            }
-            connection = startRequest();
-            payloadSerializer.serializeMetaDataNdJson(metaData);
-        }
         try {
-            writeEvent(event);
-        } catch (Exception e) {
-            // end request on error to start backing off
-            flush();
+            if (connection == null) {
+                long millisToGracePeriodEnd = gracePeriodEnd - System.currentTimeMillis();
```

Comment:
I think it would be easier to wait in `onConnectionError` that way we also don't need the `gracePeriodEnd` instance variable.

Reply:
Good idea, will change that